### PR TITLE
Allow KeyConditionExpression when model has composite key and partition key only is passed in where

### LINKED ISSUE #253: Force KeyCondition when filtering by only partition key in a table with a composite key
**Describe the bug**

I have a table with a partition key and a range key. I'm trying to use `where` to filter records by partition key only, but when i run `toDynamoDbQuery` I see that it is using FilterException instead of KeyConditionExpression.

I know that this behaviour is related to composite key definition, there is a way to force it to use KeyConditionExpression?

Thank you

**Schema**

Describe your table schema:
* entity (partition key)
* executed_at (range key)

**Debug info**

Running:

```php
dd($activityLog->where('entity', 'App\UsersPeriods#44')->toDynamoDbQuery());
```

I get:

```
BaoPham\DynamoDb\RawDynamoDbQuery {#5844
     +op: "Scan",
     +query: [
       "TableName" => "activity_logs_local",
       "FilterExpression" => "#entity = :a1",
       "ExpressionAttributeNames" => [
         "#entity" => "entity",
       ],
       "ExpressionAttributeValues" => [
         ":a1" => [
           "S" => "App\UsersPeriods#44",
         ],
       ],
     ],
   }

```

**Version info**

* Laravel: 8.83.4
* laravel-dynamodb: 6.1.3

### PR DESCRIPTION
**Pull request by thebatclaudio**
Resolve #253

### CODE REVIEW

**Comment by thebatclaudio** [comment]
Hello @baopham, have you got any news about when you will accept this PR?

**Comment by thebatclaudio** [comment]
Hello @JackPriceBurns, thank you.
I just pushed another commit

**Comment by thebatclaudio** [comment]
Ops, I broke something

**Comment by thebatclaudio** [comment]
Fixed: array_search can return 0, so I have to check if the result is === false (instead of use !array_search(...))

**Comment by thebatclaudio** [comment]
Hi @baopham and @JackPriceBurns,
sorry if I'm asking again, but have you got any news about when you will accept this PR?
How I was saying before, I need this fixes on my project and I would like to use your composer package instead of my fork on github (that is a problem for CI).

Thank you

**Comment by thebatclaudio** [comment]
Hello?

File: composer.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
-    "name": "baopham/dynamodb",
-    "description": "Eloquent syntax for DynamoDB",
+    "name": "thebatclaudio/dynamodb",
```
Comment (baopham) [inline composer.json:2]: are these changes needed here?

File: src/ConditionAnalyzer/Analyzer.php [hunk 2]
```diff
@@ -202,7 +210,7 @@ private function hasValidQueryOperator($hash, $range = null)
         $hashConditionType = $this->getCondition($hash)['type'] ?? null;
         $validQueryOp = ComparisonOperator::isValidQueryDynamoDbOperator($hashConditionType);
 
-        if ($validQueryOp && $range) {
+        if ($validQueryOp && $range && $this->getCondition($range)) {
```
Comment (baopham) [inline src/ConditionAnalyzer/Analyzer.php:213]: is this needed? should the check be in line 214 instead?

**Review by baopham — CHANGES_REQUESTED** [review]
thanks for working on this! just a couple of comments

File: composer.json (hunk 1 above)
Comment (thebatclaudio) [reply composer.json:2]: I changed this because I needed these changes, so I forked the library and republished on composer. I will remove this.

**Comment by baopham** [comment]
@nelson6e65 I think this looks good. Could you help check? if it looks good to you, will need your help to merge and publish a patch version for this PR.

File: config/dynamodb.php [hunk 3]
```diff
@@ -39,7 +39,7 @@
         ],
         'aws_iam_role' => [
             'region' => env('DYNAMODB_REGION'),
-            'debug' => true,
+            'debug' => env('DYNAMODB_DEBUG'),
```
Comment (nelson6e65) [inline config/dynamodb.php:42]:
```suggestion
            'debug' => env('DYNAMODB_DEBUG', env('APP_DEBUG', true)),
```

Maybe is a good idea to have a default value here. By default, the old one.

**Review by nelson6e65 — CHANGES_REQUESTED** [review]

File: config/dynamodb.php (hunk 3 above)
Comment (nelson6e65) [reply config/dynamodb.php:42]: Or it is `true` the default behavior?

**Comment by baopham** [comment]
> I was having trouble on setting up my local environment to test, @baopham. That's why I could not test this locally. (Maybe you can help me this Saturday on this).
> 
> Are there any failing test cases for this bug?

@nelson6e65 actually... I no longer work with PHP so I don't have anything setup unfortunately

**Review by baopham — APPROVED** [review]

**Comment by nelson6e65** [comment]
@thebatclaudio Thanks for this fix. I think that with those linted changes, this PR will be ready. 😃

File: config/dynamodb.php (hunk 3 above)
Comment (zoe-edwards) [reply config/dynamodb.php:42]: I think as it was (`env('DYNAMODB_DEBUG')`) is sufficent, otherwise it gets too messy.

File: src/ConditionAnalyzer/Analyzer.php (hunk 2 above)
Comment (nelson6e65) [reply src/ConditionAnalyzer/Analyzer.php]: > 

No, actually. It needs to be there. I tested locally. The ranged condition is valid to be Query when it is null.

So, this should be `&& $this->getCondition($range) !== null` instead

File: src/ConditionAnalyzer/Analyzer.php [hunk 4]
```diff
@@ -203,7 +211,10 @@ private function hasValidQueryOperator($hash, $range = null)
         $validQueryOp = ComparisonOperator::isValidQueryDynamoDbOperator($hashConditionType);
 
         if ($validQueryOp && $range) {
-            $rangeConditionType = $this->getCondition($range)['type'] ?? null;
+            $rangeConditionType = null;
+            if($this->getCondition($range)) {
+                $rangeConditionType = $this->getCondition($range)['type'] ?? null;
+            }
```
Comment (nelson6e65) [inline src/ConditionAnalyzer/Analyzer.php:217]: @thebatclaudio Actually, you should use the inline check as you had before. I tested locally and worked well, this new version does not work, because takes `null` as invalid operator for sort key.

Just check against `not null`: `$this->getCondition($range) !== null`, is the only condition to further check for operator type of range.

**Review by nelson6e65 — CHANGES_REQUESTED** [review]

File: src/ConditionAnalyzer/Analyzer.php (hunk 4 above)
Comment (nelson6e65) [reply src/ConditionAnalyzer/Analyzer.php:217]: Because the intention is to allow `Query` when there is no condition for range at all.

File: config/dynamodb.php (hunk 3 above)
Comment (nelson6e65) [reply config/dynamodb.php:42]: Yes, seems to work if is set to null.

File: src/ConditionAnalyzer/Analyzer.php (hunk 4 above)
Comment (thebatclaudio) [reply src/ConditionAnalyzer/Analyzer.php]: Hi @nelson6e65,
sorry, my fault. I just fixed also this.

File: config/dynamodb.php (hunk 3 above)
Comment (thebatclaudio) [reply config/dynamodb.php]: Ok, I fixed also this, but I'm reading that you decided to leave only DYNAMODB_DEBUG.
I fix this again.

Comment (thebatclaudio) [reply config/dynamodb.php:42]: Fixed!

**Review by nelson6e65 — APPROVED** [review]
Thank you. :D

LGTM and ready to be squashed and merged into master, @baopham

File: src/ConditionAnalyzer/Analyzer.php (hunk 4 above)
Comment (nelson6e65) [reply src/ConditionAnalyzer/Analyzer.php]: No worries, great job, @thebatclaudio! :D

**Comment by nelson6e65** [comment]
This is available as [v6.2.0 🚀](https://github.com/baopham/laravel-dynamodb/releases/tag/6.2.0).
